## Task 4.9 - Part 1

## Contents
#### 1. Import customers.csv
#### 2. Wrangle Data
#### 3. Complete the fundamental data quality and consistency checks
#### 4. Combine your customer data with the rest of your prepared Instacart data.

#### Question 3 - Import your analysis libraries, as well as your new customer data set as a dataframe.

In [1]:
#Import Libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [2]:
#Create path
path = r'/Users/anfran/Documents/Instacart Basket Analysis'

In [6]:
#Import data set
customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))

#### Question 4 - Wrangle the data so that it follows consistent logic; for example, rename columns with illogical names and drop columns that don’t add anything to your analysis.

In [8]:
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    194950 non-null  object
 2   Surnam        206209 non-null  object
 3   Gender        206209 non-null  object
 4   STATE         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [9]:
customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [10]:
customers.shape

(206209, 10)

In [11]:
customers.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [12]:
#Rename columns
customers = customers.rename(columns={'First Name': 'first_name','Surnam': 'last_name','Gender': 'gender', 'STATE': 'state', 'Age': 'age', 'date_joined' :'registration_date', 'n_dependants' : 'number_of_dependants', 'fam_status' : 'marital_status'})

In [13]:
#Check results
customers.head()

user_id first_name last_name  gender       state  age registration_date  \
0    26711    Deborah  Esquivel  Female    Missouri   48          1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36          1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35          1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40          1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26          1/1/2017   

   number_of_dependants marital_status  income  
0                     3        married  165665  
1                     0         single   59285  
2                     2        married   99568  
3                     0         single   42049  
4                     1        married   40374

#### Question 5 - Complete the fundamental data quality and consistency checks you’ve learned throughout this Achievement; for example, check for and address missing values and duplicates, and convert any mixed-type data.

In [14]:
#Check for null or NaN values
df_cust = customers.isnull().any(axis=1)

In [15]:
customers[df_cust].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 11259 entries, 53 to 206171
Data columns (total 10 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   user_id               11259 non-null  int64 
 1   first_name            0 non-null      object
 2   last_name             11259 non-null  object
 3   gender                11259 non-null  object
 4   state                 11259 non-null  object
 5   age                   11259 non-null  int64 
 6   registration_date     11259 non-null  object
 7   number_of_dependants  11259 non-null  int64 
 8   marital_status        11259 non-null  object
 9   income                11259 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 967.6+ KB


11259 First name values are missing. This is not an issue. 

In [16]:
#Check for duplicates
customers_dups = customers.duplicated()

In [17]:
customers_dups.info()

<class 'pandas.core.series.Series'>
RangeIndex: 206209 entries, 0 to 206208
Series name: None
Non-Null Count   Dtype
--------------   -----
206209 non-null  bool 
dtypes: bool(1)
memory usage: 201.5 KB


In [18]:
customers_dups.value_counts()

False    206209
dtype: int64

No duplicate entries found. 

In [19]:
#Check for mixed data
for col in customers.columns.tolist():
  weird = (customers[[col]].applymap(type) != customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (customers[weird]) > 0:
    print (col)

first_name


In [20]:
#Change first_name to 'str'
customers['first_name'] = customers['first_name'].astype('str')

#### Question 6 - Combine your customer data with the rest of your prepared Instacart data. (Hint: Make sure the key columns are the same data type!)

In [21]:
#Import prepared Instacart data
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined_updated_4.9.pkl'))

In [22]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0               9999999.0       196.0                1.0        0.0   
1                    15.0       196.0                1.0        1.0   
2                    21.0       196.0                1.0        1.0   
3                    29.0       196.0                1.0        1.0   
4                    28.0       196.0                1.0        1.0   

  product_name  ...    price_range_loc     busiest_day  busiest_days  \
0         Soda  ...  Mid-range product  Regularly busy  Regular days   
1         Soda  ...  Mid-range product  Regularly busy  Slowest days   
2         Soda  ...  Mid-range product  Regularly busy  Slowest days   
3         Soda  ...  Mid-range product      Least busy  Slowest days   
4         Soda  ...  Mid-range product      Least busy  Slowest days   

  busiest_period_of_day max_order  loyalty_flag average_price  Spending_flag  \
0        Average orders        10  New customer          6.37    Low spender   
1        Average orders        10  New customer          6.37    Low spender   
2           Most orders        10  New customer          6.37    Low spender   
3        Average orders        10  New customer          6.37    Low spender   
4           Most orders        10  New customer          6.37    Low spender   

  median_days_since_prior_order   order_frequency_flag  
0                          21.0  Non-frequent customer  
1                          21.0  Non-frequent customer  
2                          21.0  Non-frequent customer  
3                          21.0  Non-frequent customer  
4                          21.0  Non-frequent customer  

[5 rows x 23 columns]

In [23]:
ords_prods_merge.shape

(32404859, 23)

In [24]:
#Merge using outer join on 'user_id'
ords_prods_cust_merge = pd.merge(ords_prods_merge, customers, on = 'user_id', how = 'outer')

In [25]:
#Check results
ords_prods_cust_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0               9999999.0       196.0                1.0        0.0   
1                    15.0       196.0                1.0        1.0   
2                    21.0       196.0                1.0        1.0   
3                    29.0       196.0                1.0        1.0   
4                    28.0       196.0                1.0        1.0   

  product_name  ...   order_frequency_flag  first_name  last_name  gender  \
0         Soda  ...  Non-frequent customer       Linda     Nguyen  Female   
1         Soda  ...  Non-frequent customer       Linda     Nguyen  Female   
2         Soda  ...  Non-frequent customer       Linda     Nguyen  Female   
3         Soda  ...  Non-frequent customer       Linda     Nguyen  Female   
4         Soda  ...  Non-frequent customer       Linda     Nguyen  Female   

     state age registration_date  number_of_dependants marital_status  income  
0  Alabama  31         2/17/2019                     3        married   40423  
1  Alabama  31         2/17/2019                     3        married   40423  
2  Alabama  31         2/17/2019                     3        married   40423  
3  Alabama  31         2/17/2019                     3        married   40423  
4  Alabama  31         2/17/2019                     3        married   40423  

[5 rows x 32 columns]

In [26]:
ords_prods_cust_merge.shape

(32404859, 32)

In [ ]:
#Drop 'first_name' for customer privacy
ords_prods_cust_merge = ords_prods_cust_merge.drop(columns = ['first_name'])

In [ ]:
#Drop 'last_name' for customer privacy
ords_prods_cust_merge = ords_prods_cust_merge.drop(columns = ['last_name'])

#### Question 8 - Export this new dataframe as a pickle file so you can continue to use it in the second part of this task.

In [27]:
#Export to pickle
ords_prods_cust_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_customers_combined.pkl'))